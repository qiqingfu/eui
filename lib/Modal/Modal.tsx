import React from 'react';

export interface IProps {
  visible: boolean;
  onCancel?: React.MouseEventHandler;
  maskClosable?: boolean;
}

const Modal: React.FC<IProps> = (props: IProps) => {
  const { visible, maskClosable = true, onCancel } = props;

  const cancelHandle = (e: React.MouseEvent) => {
    if (onCancel && typeof onCancel === 'function') {
      onCancel(e);
    }
  };

  const maskClick = (e: React.MouseEvent) => {
    if (maskClosable) {
      cancelHandle(e);
    }
  };

  return visible ? (
    <>
      <div className="__e-ui-modal-mask" onClick={maskClick}></div>
      <div className="__e-ui-modal">
        <div>Header</div>
        <div>Main</div>
        <div>Footer</div>
      </div>
    </>
  ) : null;
};

export default Modal;
