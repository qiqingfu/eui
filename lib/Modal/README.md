# 开发记录

## 2021.09.07

- 创建 Modal 组件目录结构，确定部分 props 接口
- 一个必选的 visible props
- 新增 onCancel props 关闭的回调函数。范围：取消按钮、遮罩层
- 新增 maskClosable props，点击遮罩层是否关闭, 默认设置为 true
