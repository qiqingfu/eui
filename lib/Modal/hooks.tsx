import React, { useState, useEffect } from 'react';
import { Button } from '../index';
import { Type } from '../_util/index';

interface useEventReturn {
  okHandle(e: React.MouseEvent): void;
  cancelHandle(e: React.MouseEvent): void;
}

interface useFooterParams extends useEventReturn {
  footer?: React.ReactNode[] | null;
}

interface useFooterReturn {
  footerNode: React.ReactNode;
}

export const useEvent = (
  onOk?: React.MouseEventHandler,
  onCancel?: React.MouseEventHandler,
): useEventReturn => {
  const cancelHandle = (e: React.MouseEvent) => {
    if (onCancel && Type.Function(onCancel)) {
      onCancel(e);
    }
  };

  const okHandle = (e: React.MouseEvent) => {
    if (onOk && Type.Function(onOk)) {
      onOk(e);
    }
  };

  return {
    okHandle,
    cancelHandle,
  };
};

export const useFooter = ({ okHandle, cancelHandle, footer }: useFooterParams): useFooterReturn => {
  const initFooter = (
    <div className="e-modal__footer">
      <Button key="cancel" size="small" onClick={cancelHandle}>
        取消
      </Button>
      <Button key="confirm" size="small" type="primary" onClick={okHandle}>
        确认
      </Button>
    </div>
  );

  const [footerNode, setFooter] = useState<React.ReactNode | null>(initFooter);
  // 后面需要进行优化
  useEffect(() => {
    if (footer === null) {
      setFooter(null);
    } else if (footer === undefined) {
      setFooter(initFooter);
    } else {
      setFooter(<div className="e-modal__footer">{footer}</div>);
    }
  }, [footer]);

  return {
    footerNode,
  };
};
