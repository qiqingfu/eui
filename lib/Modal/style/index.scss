.e-modal {
  @include pos-center;
  z-index: $z-mask;
  width: 500px;
  background-color: #fff;
  border-radius: 2px;
  box-shadow: 0 3px 6px -4px #0000001f, 0 6px 16px #00000014, 0 9px 28px 8px #0000000d;

  &__header {
    padding: 16px 24px;
    border-bottom: solid 1px #f0f0f0;
  }

  &__close {
    position: absolute;
    right: 0;
    top: 0;
    width: 56px;
    height: 56px;
    line-height: 56px;
    text-align: center;
    cursor: pointer;
  }

  &__main {
    padding: 24px;
    font-size: 14px;
    word-wrap: break-word;
  }

  &__footer {
    padding: 10px;
    text-align: right;
    border-top: solid 1px #f0f0f0;
  }

  &__mask {
    position: fixed;
    left: 0;
    top: 0;
    right: 0;
    bottom: 0;
    z-index: $z-mask;
    background-color: #00000073;
  }
}
