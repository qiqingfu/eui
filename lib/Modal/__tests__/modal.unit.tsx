import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Modal from '../index';

describe('Modal Component', () => {
  it('Modal renders correctly', () => {
    const tree = renderer.create(<Modal visible></Modal>);
    expect(tree).toMatchSnapshot();
  });

  it('Modal props.visible', () => {
    const wrapper = shallow(<Modal visible></Modal>);
    expect(wrapper.find('.e-modal').exists()).toBeTruthy();
    wrapper.setProps({ visible: false });
    expect(wrapper.find('.e-modal').exists()).toBeFalsy();
  });

  it('Modal props.maskClosable set false', () => {
    const fn = jest.fn();
    const wrapper = shallow(<Modal visible onCancel={fn}></Modal>);
    wrapper.find('.e-modal__mask').simulate('click');
    expect(fn).toBeCalledTimes(1);
    wrapper.setProps({ maskClosable: false });
    wrapper.find('.e-modal__mask').simulate('click');
    expect(fn).toBeCalledTimes(1);
  });

  describe('Modal Event', () => {
    it('Click cancel button call props onCancel', () => {
      const fn = jest.fn();
      const wrapper = shallow(<Modal visible onCancel={fn}></Modal>);
      wrapper.find('.e-modal__footer').childAt(0).simulate('click');
      expect(fn).toBeCalledTimes(1);
    });

    it('Click ok button call props onOk', () => {
      const fn = jest.fn();
      const wrapper = shallow(<Modal visible onOk={fn}></Modal>);
      wrapper.find('.e-modal__footer').childAt(1).simulate('click');
      expect(fn).toBeCalledTimes(1);
    });
  });

  it('Modal props.closable', () => {
    // TODO
  });
});
