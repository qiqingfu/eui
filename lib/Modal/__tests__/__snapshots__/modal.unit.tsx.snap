// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component Modal renders correctly 1`] = `
Array [
  <div
    className="e-modal__mask"
    onClick={[Function]}
  />,
  <div
    className="e-modal"
  >
    <div
      className="e-modal__header"
    >
      Header
      <span
        className="e-modal__close"
        onClick={[Function]}
      >
        <svg
          className="e-ui-icon"
          style={Object {}}
        >
          <use
            xlinkHref="#close"
          />
        </svg>
      </span>
    </div>
    <div
      className="e-modal__main"
    >
      Main
    </div>
    <div
      className="e-modal__footer"
    >
      <button
        className="e-btn small"
        onClick={[Function]}
      >
        <span>
          取消
        </span>
      </button>
      <button
        className="e-btn small"
        onClick={[Function]}
      >
        <span>
          确认
        </span>
      </button>
    </div>
  </div>,
]
`;
